File: test files/typescript/senior_avancado.ts
type TaskStatus = 'todo' | 'doing' | 'done' | 'blocked';
type TaskTag = 'critical' | 'customer-facing' | 'billing-sensitive';

type CompleteTaskFailure =
  | 'TASK_NOT_FOUND'
  | 'TASK_WITHOUT_OWNER'
  | 'TASK_BLOCKED'
  | 'INVALID_ESTIMATE'
  | 'VERSION_CONFLICT'
  | 'AUDIT_UNAVAILABLE';

type Result<T, E extends string> =
  | { ok: true; data: T }
  | { ok: false; error: E };

type TaskId = number & { readonly __brand: 'TaskId' };

type CompletionCommand = Readonly<{
  taskId: TaskId;
  expectedVersion: number;
  actorId: number;
}>;

type TaskSnapshot = Readonly<{
  id: TaskId;
  title: string;
  ownerId: number | null;
  status: TaskStatus;
  estimateHours: number;
  version: number;
  tags: ReadonlyArray<TaskTag>;
}>;

type DomainEvent = Readonly<{
  type: 'task.completed';
  entityId: TaskId;
  occurredAt: string;
  actorId: number;
}>;

interface TaskRepository {
  findById(id: TaskId): Promise<TaskSnapshot | null>;
  save(task: TaskSnapshot): Promise<TaskSnapshot>;
}

interface AuditGateway {
  publish(entry: AuditEntry): Promise<void>;
}

interface DomainEventBus {
  publish(event: DomainEvent): Promise<void>;
}

interface MonitoringGateway {
  info(event: string, context: Record<string, unknown>): void;
  warn(event: string, context: Record<string, unknown>): void;
}

interface Clock {
  now(): Date;
}

type AuditEntry = {
  entityId: TaskId;
  action: 'task.completed';
  at: string;
  metadata: {
    actorId: number;
    previousStatus: TaskStatus;
    estimateHours: number;
    previousVersion: number;
    tags: ReadonlyArray<TaskTag>;
  };
};

class TaskCompletionPolicy {
  evaluate(
    task: TaskSnapshot
  ): Result<TaskSnapshot, Extract<CompleteTaskFailure, 'TASK_WITHOUT_OWNER' | 'TASK_BLOCKED' | 'INVALID_ESTIMATE'>> {
    if (task.ownerId === null) {
      return { ok: false, error: 'TASK_WITHOUT_OWNER' };
    }

    if (task.status === 'blocked') {
      return { ok: false, error: 'TASK_BLOCKED' };
    }

    if (task.estimateHours <= 0) {
      return { ok: false, error: 'INVALID_ESTIMATE' };
    }

    return { ok: true, data: task };
  }
}

class CompleteTaskUseCase {
  constructor(
    private readonly taskRepository: TaskRepository,
    private readonly auditGateway: AuditGateway,
    private readonly domainEventBus: DomainEventBus,
    private readonly monitoringGateway: MonitoringGateway,
    private readonly clock: Clock,
    private readonly policy: TaskCompletionPolicy
  ) {}

  async execute(
    command: CompletionCommand
  ): Promise<Result<{ taskId: TaskId; completedAt: string; nextVersion: number }, CompleteTaskFailure>> {
    const task = await this.taskRepository.findById(command.taskId);
    if (!task) {
      this.monitoringGateway.warn('task.complete.not_found', {
        taskId: command.taskId,
        actorId: command.actorId,
      });
      return { ok: false, error: 'TASK_NOT_FOUND' };
    }

    const validation = this.policy.evaluate(task);
    if (!validation.ok) {
      this.monitoringGateway.warn('task.complete.validation_failed', {
        taskId: command.taskId,
        actorId: command.actorId,
        code: validation.error,
        tags: task.tags,
      });
      return validation;
    }

    if (task.version !== command.expectedVersion) {
      this.monitoringGateway.warn('task.complete.version_conflict', {
        taskId: task.id,
        actorId: command.actorId,
        expectedVersion: command.expectedVersion,
        currentVersion: task.version,
        tags: task.tags,
      });
      return { ok: false, error: 'VERSION_CONFLICT' };
    }

    const completedAt = this.clock.now().toISOString();
    const updatedTask: TaskSnapshot = {
      ...task,
      status: 'done',
      version: task.version + 1,
    };

    try {
      await this.taskRepository.save(updatedTask);
      await this.auditGateway.publish({
        entityId: updatedTask.id,
        action: 'task.completed',
        at: completedAt,
        metadata: {
          actorId: command.actorId,
          previousStatus: task.status,
          estimateHours: task.estimateHours,
          previousVersion: task.version,
          tags: task.tags,
        },
      });
      await this.domainEventBus.publish({
        type: 'task.completed',
        entityId: updatedTask.id,
        occurredAt: completedAt,
        actorId: command.actorId,
      });
    } catch {
      this.monitoringGateway.warn('task.complete.side_effect_unavailable', {
        taskId: task.id,
        actorId: command.actorId,
        previousVersion: task.version,
        tags: task.tags,
      });
      return { ok: false, error: 'AUDIT_UNAVAILABLE' };
    }

    this.monitoringGateway.info('task.complete.succeeded', {
      taskId: updatedTask.id,
      actorId: command.actorId,
      nextVersion: updatedTask.version,
      tags: updatedTask.tags,
    });

    return {
      ok: true,
      data: {
        taskId: updatedTask.id,
        completedAt,
        nextVersion: updatedTask.version,
      },
    };
  }
}

File: test files/java/senior_avancado.java
import java.time.Clock;
import java.util.Map;
import java.util.Optional;

public class SeniorAvancado {
    enum TaskStatus {
        TODO, DOING, DONE, BLOCKED
    }

    record TaskSnapshot(int id, String title, Integer ownerId, TaskStatus status, int estimateHours, int version) {}

    record AuditMetadata(TaskStatus previousStatus, int estimateHours, int previousVersion) {}

    record AuditEntry(int entityId, String action, String at, AuditMetadata metadata) {}

    record CompleteTaskCommand(int taskId) {}

    record CompleteTaskOutput(int taskId, String completedAt, int nextVersion) {}

    sealed interface CompleteTaskFailure permits TaskNotFound, ValidationFailure, ExternalDependencyFailure {
        String code();
    }

    record TaskNotFound() implements CompleteTaskFailure {
        public String code() {
            return "TASK_NOT_FOUND";
        }
    }

    record ValidationFailure(String code) implements CompleteTaskFailure {}

    record ExternalDependencyFailure(String code) implements CompleteTaskFailure {}

    sealed interface Result<T> permits Success, Failure {}

    record Success<T>(T value) implements Result<T> {}

    record Failure<T>(CompleteTaskFailure error) implements Result<T> {}

    interface TaskRepository {
        Optional<TaskSnapshot> findById(int id);
        TaskSnapshot save(TaskSnapshot task);
    }

    interface AuditGateway {
        void publish(AuditEntry entry);
    }

    interface MonitoringGateway {
        void info(String event, Map<String, Object> context);
        void warn(String event, Map<String, Object> context);
    }

    static class TaskCompletionPolicy {
        Optional<ValidationFailure> validate(TaskSnapshot task) {
            if (task.ownerId() == null) {
                return Optional.of(new ValidationFailure("TASK_WITHOUT_OWNER"));
            }

            if (task.status() == TaskStatus.BLOCKED) {
                return Optional.of(new ValidationFailure("TASK_BLOCKED"));
            }

            if (task.estimateHours() <= 0) {
                return Optional.of(new ValidationFailure("INVALID_ESTIMATE"));
            }

            return Optional.empty();
        }
    }

    static class CompleteTaskUseCase {
        private final TaskRepository taskRepository;
        private final AuditGateway auditGateway;
        private final MonitoringGateway monitoringGateway;
        private final Clock clock;
        private final TaskCompletionPolicy taskCompletionPolicy;

        CompleteTaskUseCase(
            TaskRepository taskRepository,
            AuditGateway auditGateway,
            MonitoringGateway monitoringGateway,
            Clock clock,
            TaskCompletionPolicy taskCompletionPolicy
        ) {
            this.taskRepository = taskRepository;
            this.auditGateway = auditGateway;
            this.monitoringGateway = monitoringGateway;
            this.clock = clock;
            this.taskCompletionPolicy = taskCompletionPolicy;
        }

        Result<CompleteTaskOutput> execute(CompleteTaskCommand command) {
            Optional<TaskSnapshot> maybeTask = taskRepository.findById(command.taskId());
            if (maybeTask.isEmpty()) {
                monitoringGateway.warn("task.complete.not_found", Map.of("taskId", command.taskId()));
                return new Failure<>(new TaskNotFound());
            }

            TaskSnapshot task = maybeTask.get();
            Optional<ValidationFailure> validationFailure = taskCompletionPolicy.validate(task);
            if (validationFailure.isPresent()) {
                monitoringGateway.warn("task.complete.validation_failed", Map.of(
                    "taskId", task.id(),
                    "code", validationFailure.get().code()
                ));
                return new Failure<>(validationFailure.get());
            }

            String completedAt = clock.instant().toString();
            TaskSnapshot updatedTask = new TaskSnapshot(
                task.id(),
                task.title(),
                task.ownerId(),
                TaskStatus.DONE,
                task.estimateHours(),
                task.version() + 1
            );

            try {
                taskRepository.save(updatedTask);
                auditGateway.publish(new AuditEntry(
                    updatedTask.id(),
                    "task.completed",
                    completedAt,
                    new AuditMetadata(task.status(), task.estimateHours(), task.version())
                ));
            } catch (RuntimeException exception) {
                monitoringGateway.warn("task.complete.audit_unavailable", Map.of(
                    "taskId", task.id(),
                    "previousVersion", task.version()
                ));
                return new Failure<>(new ExternalDependencyFailure("AUDIT_UNAVAILABLE"));
            }

            monitoringGateway.info("task.complete.succeeded", Map.of(
                "taskId", updatedTask.id(),
                "nextVersion", updatedTask.version()
            ));

            return new Success<>(new CompleteTaskOutput(updatedTask.id(), completedAt, updatedTask.version()));
        }
    }
}
